### PR TITLE
Content fields indexing

### PR DESCRIPTION
**Pull request by Skrypt**
Work in progress

image

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/BooleanFieldIndexProvider.cs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+using OrchardCore.Data;
+using YesSql.Indexes;
+
+namespace OrchardCore.ContentFields.Indexing
+{
+    // Remark: 
+
+    public class BooleanFieldIndex : MapIndex
+    {
+        public string FieldName { get; set; }
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/BooleanFieldIndexProvider.cs]: Like in Taxonomies you need the ContentType and the ContentPart properties on all indices

**Comment by sebastienros** [comment]
You can use a base `MapIndex` class that would contain the ContentType, Part and field names. You can also use a base class for the provider that would handle the types/fields filtering, and just have to override how to build the index.

**Comment by Skrypt** [comment]
Ok all good ideas

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/ContentFieldIndexProvider.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        private IContentDefinitionManager _contentDefinitionManager;
+
+        public ContentFieldIndexProvider(IServiceProvider serviceProvider)
+        {
+            _serviceProvider = serviceProvider;
+        }
+
+        public override void Describe(DescribeContext<ContentItem> context)
+        {
+            context.For<ContentFieldIndex>()
```
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/ContentFieldIndexProvider.cs]: I still need to override that method but this could be all removed unless we want to have a ContentFieldIndex table added in the migration.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/ContentPickerFieldIndexProvider.cs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+using OrchardCore.ContentManagement;
+using OrchardCore.ContentManagement.Metadata;
+using OrchardCore.Data;
+using YesSql.Indexes;
+
+namespace OrchardCore.ContentFields.Indexing
+{
+    // Remark: 
+
+    public class ContentPickerFieldIndex : ContentFieldIndex
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/ContentPickerFieldIndexProvider.cs]: Look at the TaxonomiesField, it is a list of ContentItemId instead so you can filter them. You might no be able to inherit from the same base class.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/ContentFieldIndexProvider.cs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+                        var jField = (JObject)jPart[fieldDefinition.Name];
+
+                        if (jField == null)
+                        {
+                            continue;
+                        }
+
+                        var field = jField.ToObject<TextField>();
+
+                        results.Add(new ContentFieldIndex
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/ContentFieldIndexProvider.cs]: Create an absctract `CreateIndex` that returns a generic type defined on this class. This way the sub-classes can only implement this new method that return the index object only. Lots of code to remove from all providers then. If not clear ping me and we'll pair program.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/ContentPickerFieldIndexProvider.cs (hunk 3 above)
Comment (Skrypt) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/ContentPickerFieldIndexProvider.cs]: Ok you want one row per ContentItemId. Though my solution would still work if using a LIKE '%ContentItemIdString%'

**Comment by Skrypt** [comment]
I think it was ready but I need to make a double check on the abstract classes that we're requested. I need to check if I committed these.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/ContentFieldIndexProvider.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+namespace OrchardCore.ContentFields.Indexing
+{
+    public abstract class ContentFieldIndex : MapIndex
+    {
+        public string ContentType { get; set; }
+        public string ContentPart { get; set; }
+        public string ContentField { get; set; }
+    }
+
+    public abstract class ContentFieldIndexProvider : IndexProvider<ContentItem>, IScopedIndexProvider
```
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/ContentFieldIndexProvider.cs]: Here is the abstract class

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/ContentFieldIndexProvider.cs [hunk 6]
```diff
@@ -0,0 +1,26 @@
+using System;
+using System.Collections.Generic;
+using OrchardCore.ContentManagement;
+using OrchardCore.Data;
+using YesSql.Indexes;
+
+namespace OrchardCore.ContentFields.Indexing
+{
+    public abstract class ContentFieldIndex : MapIndex
```
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/ContentFieldIndexProvider.cs]: Here is the abstract class

**Comment by sebastienros** [comment]
Note for myself, we need two other columns (Publish, Latest) so we can store records for drafts, and provide an easy way to query Published and Draft items.

File: src/OrchardCore.Modules/OrchardCore.Contents/Recipes/ContentStep.cs [hunk 7]
```diff
@@ -49,6 +49,8 @@ public async Task ExecuteAsync(RecipeExecutionContext context)
                     _session.Save(existing);
                 }
             }
+
+            await _session.CommitAsync();
```
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.Contents/Recipes/ContentStep.cs]: I think that you don't need this line anymore if you merge the dev branch where we now commit the session automatically before disposing the shell scope.

Comment (Skrypt) [reply src/OrchardCore.Modules/OrchardCore.Contents/Recipes/ContentStep.cs]: I can remove it but I don't think it will cause any issue calling it early right?

**Comment by Skrypt** [comment]
I applied the required changes on this PR.

image

File: src/OrchardCore.Modules/OrchardCore.Contents/Recipes/ContentStep.cs (hunk 7 above)
Comment (Skrypt) [reply src/OrchardCore.Modules/OrchardCore.Contents/Recipes/ContentStep.cs]: BTW I don't think this PR has anything to do with the ContentStep recipes.

**Comment by Skrypt** [comment]
When removing a content item it will not set the published field to false. That's an issue I found so I'm keeping it as WIP.

File: src/OrchardCore.Modules/OrchardCore.Contents/Recipes/ContentStep.cs (hunk 7 above)
Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.Contents/Recipes/ContentStep.cs]: > BTW I don't think this PR has anything to do with the ContentStep recipes.

Yes not directly but the issue was to do CommitAsync in session Dispose that may use an already disposed scoped service, so it may happen in different contexts and was happening during a setup recipe.

> I can remove it but I don't think it will cause any issue calling it early right?

Yes but maybe worth to re-try some setups without this line and i think that now the right pattern is to let the session being committed automatically (and once) just before the shell scope is disposed, this after any session save that may occur here or after.

Comment (Skrypt) [reply src/OrchardCore.Modules/OrchardCore.Contents/Recipes/ContentStep.cs]: Let's open an issue and a PR to keep track of that specific issue.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Migrations.cs [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+using OrchardCore.ContentFields.Indexing;
+
+namespace OrchardCore.ContentFields
+{
+    public class Migrations : DataMigration
+    {
+        public int Create()
+        {
+            SchemaBuilder.CreateMapIndexTable(nameof(TextFieldIndex), table => table
+                .Column<string>("ContentType", column => column.WithLength(255))
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Migrations.cs]: This has to match the values from the other tables.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Migrations.cs [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+                .Column<string>("ContentPart", column => column.WithLength(255))
+                .Column<string>("ContentField", column => column.WithLength(255))
+                .Column<bool>("Published")
+                .Column<bool>("Latest")
+                .Column<string>("Text", column => column.Nullable().WithLength(4000))
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Migrations.cs]: Ensure we use the same name as in the field property.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Migrations.cs [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+                .Column<string>("RichText", column => column.Nullable().Unlimited())
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Migrations.cs]: `BigText`

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Migrations.cs [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+            SchemaBuilder.CreateMapIndexTable(nameof(DateFieldIndex), table => table
+                .Column<string>("ContentType", column => column.WithLength(255))
+                .Column<string>("ContentPart", column => column.WithLength(255))
+                .Column<string>("ContentField", column => column.WithLength(255))
+                .Column<bool>("Published")
+                .Column<bool>("Latest")
+                .Column<DateTime>("Date", column => column.Nullable())
+            );
+
+            SchemaBuilder.CreateMapIndexTable(nameof(ContentPickerFieldIndex), table => table
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Migrations.cs]: Please add database indices for each of the columns

File: src/OrchardCore.Modules/OrchardCore.Contents/Recipes/ContentStep.cs (hunk 7 above)
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Contents/Recipes/ContentStep.cs]: Worst case, if you need something it should be a flush, not a commit.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Migrations.cs (hunk 8 above)
Comment (Skrypt) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Migrations.cs]: Changed to `.Column<string>("ContentType", column => column.WithLength(ContentItemIndex.MaxContentTypeSize))`

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Migrations.cs (hunk 10 above)
Comment (Skrypt) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Migrations.cs]: done

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Migrations.cs (hunk 11 above)
Comment (Skrypt) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Migrations.cs]: done

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Migrations.cs (hunk 8 above)
Comment (Skrypt) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Migrations.cs]: Let me know if we should also use it on ContentPart and ContentField since they are also both GUID's

Comment (sebastienros) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Migrations.cs]: What is a Guid also?

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/TextFieldIndexProvider.cs [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+                        if (jField == null)
+                        {
+                            continue;
+                        }
+
+                        var field = jField.ToObject<TextField>();
+
+                        if (!String.IsNullOrEmpty(field.Text))
+                        {
+                            if (field.Text.Length > 4000)
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/TextFieldIndexProvider.cs]: I think we should always set the value in `BigText` and always take the first 4000 chars in `Text`. This way in a query you can always use `Text` to filter and always use `BigText` to retrieve the full value.
Also only `Text` should be indexed.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Migrations.cs [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+
+            SchemaBuilder.AlterTable(nameof(TextFieldIndex), table => table
+                .CreateIndex("IDX_TextFieldIndex_Latest", "Latest")
+            );
+
+            SchemaBuilder.AlterTable(nameof(TextFieldIndex), table => table
+                .CreateIndex("IDX_TextFieldIndex_Text", "Text")
+            );
+
+            SchemaBuilder.AlterTable(nameof(TextFieldIndex), table => table
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Migrations.cs]: Don't index `BigText`

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Migrations.cs (hunk 8 above)
Comment (Skrypt) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Migrations.cs]: They are unique random strings ... maybe not guid or uuid ... but something 😄

Comment (Skrypt) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Migrations.cs]: I think we should only care about the lenght of them though 😉

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Migrations.cs (hunk 13 above)
Comment (Skrypt) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Migrations.cs]: done

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/TextFieldIndexProvider.cs (hunk 12 above)
Comment (Skrypt) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/TextFieldIndexProvider.cs]: done

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Migrations.cs (hunk 8 above)
Comment (Skrypt) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Migrations.cs]: I think we should maybe eventually add the ContentPart and ContentField keys to the IndexingConstants.cs that's what I meant.

Comment (Skrypt) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Migrations.cs]: And also add something for MaxContentPartSize, MaxContentFieldSize

**Comment by jnucpz** [comment]
Question: if i filter two or multipule field , how to solve?  this solution just fit only one field.

**Comment by sebastienros** [comment]
@jnucpz that's a very good point, I think we already talked about it at some point, but this is the simplest solution we can find right now. A better approach would be to create custom tables for each type of query we want to execute, somehow like YesSql or RavendDb indices. Or reuse what the indexing module is providing and map it to tables automatically.

At least right now you can query two fields if they are not of the same type. And there is probably some SQL magic to do that would allow to query two different row, like a union.

**Comment by Skrypt** [comment]
This should be ready too

**Comment by jnucpz** [comment]
@Skrypt @sebastienros  I want to create custom index table for every content type, the type content  map to  the same name index tables。the problem is, i need to dynamic create type index,and need to change the for<> method and query method. it is complex. 

 public override void Describe(DescribeContext<ContentItem> context)
        {
            context.For<**ContainedPartIndex**>()
                .Map(contentItem =>

what is your good sugestion? how i should think ?

**Comment by Skrypt** [comment]
```C#
    public class ContentTypeIndexProvider : IndexProvider<ContentItem>, IScopedIndexProvider
    {
        private readonly IShellHost _shellHost;
        private readonly ShellSettings _shellSettings;
        private IContentManager _contentManager;

        public ContentTypeIndexProvider(
            IShellHost shellHost,
            ShellSettings shellSettings)
        {
            _shellHost = shellHost;
            _shellSettings = shellSettings;
        }

        public override void Describe(DescribeContext<ContentItem> context)
        {
            context.For<ContainedPartIndex>()
                .Map(contentItem =>
                {
                    // Create a scope for the content manager
                    using (var scope = _shellHost.GetScopeAsync(_shellSettings))
                    {
                        _contentManager = scope.Result.ServiceProvider.GetRequiredService<IContentManager>();

                        if (contentItem.ContentType == "MyContentType")
                        {
```

You will need one context.For per contentType. This kind of indexing might be risky as a content type definition could change over time and it would require to be able to change the index table schema everytime. We use migrations for that matter normally to keep a track of the changes made to the DB but here you would require to make changes on runtime. This is mainly why we did not do such thing in OC so far and I'm skeptical on the need for indexing every content type. I do it for some content types but the Lucene indexing changes I made recently made this unnecessary.

**Comment by Skrypt** [comment]
> which are not Published or Latest
I'm testing and I think you meant : not Published AND not Latest should not be indexed.

**Comment by sebastienros** [comment]
Yep, what I meant

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Startup.cs [hunk 14]
```diff
@@ -105,6 +107,18 @@ public override void ConfigureServices(IServiceCollection services)
...
+            services.AddScoped<IScopedIndexProvider, TextFieldIndexProvider>();
+            services.AddScoped<IScopedIndexProvider, BooleanFieldIndexProvider>();
+            services.AddScoped<IScopedIndexProvider, NumericFieldIndexProvider>();
+            services.AddScoped<IScopedIndexProvider, DateTimeFieldIndexProvider>();
+            services.AddScoped<IScopedIndexProvider, DateFieldIndexProvider>();
+            services.AddScoped<IScopedIndexProvider, ContentPickerFieldIndexProvider>();
+            services.AddScoped<IScopedIndexProvider, TimeFieldIndexProvider>();
+            services.AddScoped<IScopedIndexProvider, LinkFieldIndexProvider>();
+            services.AddScoped<IScopedIndexProvider, HtmlFieldIndexProvider>();
+            services.AddScoped<IScopedIndexProvider, YoutubeFieldIndexProvider>();
```
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Startup.cs]: @sebastienros Let me know if you want me to remove that one.

Comment (sebastienros) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Startup.cs]: Yes, don't do the YT one. We can create a video editor/display for the MEdia field instead. I need to find the PR where we said what we should do instead of this one.

**Comment by Skrypt** [comment]
Ok, I think it covers pretty much what I could think of so far. Maybe missing unit tests. Let me know if I'm missed something.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/LinkFieldIndexProvider.cs [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+                        if (jField == null)
+                        {
+                            continue;
+                        }
+
+                        var field = jField.ToObject<LinkField>();
+
+                        if (!String.IsNullOrEmpty(field.Text))
+                        {
+                            results.Add(new LinkFieldIndex
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/LinkFieldIndexProvider.cs]: Looks like you could create an abstract base class as they are all the same.
Also aren't you missing the test that `!Published && !Latest` should return null?

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/LinkFieldIndexProvider.cs [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+                {
+                    // Can we safely ignore this content item?
+                    if (_ignoredTypes.Contains(contentItem.ContentType))
+                    {
+                        return null;
+                    }
+
+                    if (!contentItem.Latest && !contentItem.Published)
+                    {
+                        return null;
```
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/LinkFieldIndexProvider.cs]: It's here

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/TextFieldIndexProvider.cs [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+                        var jField = (JObject)jPart[fieldDefinition.Name];
+
+                        if (jField == null)
+                        {
+                            continue;
+                        }
+
+                        var field = jField.ToObject<TextField>();
+
+                        if (!String.IsNullOrEmpty(field.Text))
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/TextFieldIndexProvider.cs]: Why? Wouldn't we want to search for an empty string?

Comment (Skrypt) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/TextFieldIndexProvider.cs]: You are right

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/LinkFieldIndexProvider.cs [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+            _serviceProvider = serviceProvider;
+        }
+
+        public override void Describe(DescribeContext<ContentItem> context)
+        {
+            context.For<LinkFieldIndex>()
+                .Map(contentItem =>
+                {
+                    // Can we safely ignore this content item?
+                    if (_ignoredTypes.Contains(contentItem.ContentType))
```
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/LinkFieldIndexProvider.cs]: This is a relic from 01 that I kept. I'm not sure of it's value.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/LinkFieldIndexProvider.cs [hunk 19]
```diff
@@ -72,21 +72,18 @@ public override void Describe(DescribeContext<ContentItem> context)
...
-                            });
-                        }
+                            Latest = contentItem.Latest,
+                            Published = contentItem.Published,
+                            ContentItemId = contentItem.ContentItemId,
+                            ContentItemVersionId = contentItem.ContentItemVersionId,
+                            ContentType = contentItem.ContentType,
+                            ContentPart = fieldDefinition.PartDefinition.Name,
+                            ContentField = fieldDefinition.Name,
+                            Url = field.Url.Substring(0, Math.Min(field.Url.Length, 4000)),
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/LinkFieldIndexProvider.cs]: You need to test for nullability in this case

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/LinkFieldIndexProvider.cs (hunk 15 above)
Comment (Skrypt) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/LinkFieldIndexProvider.cs]: Ok working on it but having issues making the base.Describe work.

Comment (Skrypt) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/LinkFieldIndexProvider.cs]: The idea is that in a Describe we do context.For<BooleanFieldIndex>() for example.
But if I want to pass a dynamic XFieldIndex type for the base.Describe(context) then I'm stuck there 😄

Comment (Skrypt) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/LinkFieldIndexProvider.cs]: And if I do context.For<ContentFieldIndex>() then it would require that I have an other db table of that name and that's not the goal.

**Comment by sebastienros** [comment]
Just one more thing ;)

Make it a separate feature, and add a md file that list all the tables and their fields. Or there is no way people will be able to query them.
